challenge_find_my_mates/src/find_my_mates.py

```diff
@@ -0,0 +1,338 @@
+        self._search_timeout = search_timeout
+        self._look_distance = look_distance
+        self._min_dist = min_dist
+        self._people = []
+
+    def execute(self, userdata=None):
+        rospy.loginfo("Trying to find my mates")
+        self._robot.head.look_at_standing_person()
+        self._robot.speech.speak("Please look at me while I am looking for you",
+                                 block=False)
+
+        start_time = rospy.Time.now()
+
+        look_distance = 2.0  # magic number 4
+        look_angles = [f * math.pi / d if d != 0 else 0.0 for f in [-1, 1] for d in [0, 6, 4, 2.3]]  # Magic numbers
+        head_goals = [kdl_conversions.VectorStamped(x=look_distance * math.cos(angle),
+                                                    y=look_distance * math.sin(angle), z=1.7,
+                                                    frame_id="/%s/base_link" % self._robot.robot_name)
+                      for angle in look_angles]
+        head_goals.extend([kdl_conversions.VectorStamped(x=2.0 * look_distance * math.cos(angle),
+                                                         y=2.0 * look_distance * math.sin(angle), z=1.7,
+                                                         frame_id="/%s/base_link" % self._robot.robot_name)
+                           for angle in look_angles])
+        n = 0 # This represents the number of detected people, once four people are detected, this state is exited. 
+        i = 0
+
+        while (rospy.Time.now() - start_time).to_sec() < self._search_timeout:
+            if self.preempt_requested():
+                return 'Aborted'
+
+            self._robot.head.look_at_point(head_goals[i])
+            i += 1
+            if i == len(head_goals):
+                i = 0
+            self._robot.head.wait_for_motion_done()
+            raw_detections = self._robot.perception.detect_faces()
+
+            # if not raw_detections:
+            #     continue
+            if raw_detections:
+                for detection in raw_detections:
+                    roi = detection.roi
+                    try:
+                        person_pos_kdl = self._robot.perception.project_roi(roi=roi, frame_id="map")
+                    except Exception as e:
+                        rospy.logerr("head.project_roi failed: %s", e)
+                        continue  # move onto next detection
+
+                    detected_person = self._robot.ed.get_closest_laser_entity(radius=self._look_distance,
+                                                                              center_point=person_pos_kdl)
+
+                    room_entity = self._robot.ed.get_entity(id=ROOM_ID)
+                    if not room_entity.in_volume(detected_person.pose.extractVectorStamped(), 'in'):
+                        detected_person = None
+
+                    for person in self._people:
+                        if person.pose.extractVectorStamped() - detected_person._pose.extractVectorStamped() > \
```

Comment:
This if looks very complex.

Reply:
It is, but it's also necessary, what solution would you propose?

Reply:
Now I also see the mixed usage of `.pose` and `._pose`.

But create some extra variables, that would already clear things up.
`get_entities` can return multiple entities, so this code can break.
There is a repeating element in it, so that could be put in a function
Add some comments about what is happening here.

Reply:
I created extra variables with names that clarify what's going on. ```get_entities``` could indeed break, but if it would break on the starting point it will already break at the start first I suppose. I fear attempting to catch that possible error will over-complicate this. I fixed the mixed use of ```pose ``` and ```_pose```.